### PR TITLE
[RW-1851][Risk=low] Account Creation Modals React conversion 

### PR DESCRIPTION
**Pull request by UrsaStutsman**
May have jumped the gun requesting reviews - though Brett's is super helpful, so maybe not. Anyways - this PR will be changing as Scout and I decided this will be the "counter-proposal" to their "styles as js objects" proposal and I'll be incorporating SCSS instead. We think this is a good idea particularly since I've never used SCSS before, and it will give us a good idea as to how "easy" it is for previous non-users to pick up.

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     "codegen": "yarn run merge-swagger && yarn run codegen-clean && yarn run codegen-swagger && yarn run codegen-swagger-fetch && yarn run post-codegen && yarn run notebooks-codegen-swagger && yarn run notebooks-post-codegen",
     "merge-swagger": "../api/build.rb merge-yaml",
     "codegen-clean": "rm -rf src/generated src/notebooks-generated",
+    "codegen-swagger-fetch": "java -jar ../aou-utils/swagger-codegen-cli.jar generate --lang typescript-fetch --input-spec ../api/src/main/resources/merged.yaml --output src/generated/fetch",
```
Comment (panentheos) [inline ui/package.json]: This looks like a duplicate line.

File: ui/src/app/react-components/modals.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    justifyContent: 'flex-end' as 'flex-end',
+    marginBottom: '.4rem',
+    marginRight: '.4rem',
+    marginTop: '.4rem'
+  }
+};
+
+export const Modal = ({children, style = {}, ...props}) => {
+  return <div><div style={{...styles.modalBackdrop}}></div>
+    <div {...props} style={{...styles.modalMain, ...style}}>{children}</div></div>;
```
Comment (panentheos) [inline ui/src/app/react-components/modals.tsx]: This is a good starting point, but as a follow-on we should probably bring in `react-modal`, which handles a lot of the fiddly bits, including things like keyboard shortcuts, focus trapping, etc. We'll still be able to completely customize the appearance and animations.

File: ui/src/app/react-components/modals.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+};
+export const ModalTitle = ({children, style = {}, ...props}) =>
+  <div {...props} style={{...styles.modalTitle, ...style}}>{children}</div>;
```
Comment (panentheos) [inline ui/src/app/react-components/modals.tsx]: FYI: Although children has a special syntax in JSX, it otherwise acts like a normal prop, so you could omit it entirely here if you wanted, and it would get passed down implicitly via the `...props`

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+function isBlank(s: string) {
+  return (!s || /^\s*$/.test(s));
+}
+
+class AccountCreationModalsReact extends React.Component<any, any> {
+
+  state: {
+    changingEmail: boolean,
+    resendingEmail: boolean,
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: These two pieces of state look like they are never used. Can they be removed?

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+      return;
+    }
+    this.props.passNewEmail(this.state.contactEmail);
+    const request: UpdateContactEmailRequest = {
+      username: this.props.username,
+      contactEmail: this.state.contactEmail,
+      creationNonce: this.props.creationNonce
+    };
+    const args: FetchArgs = ProfileApiFetchParamCreator().updateContactEmail(request);
+    fetch(this.fullUrl(args.url), args.options).then(() => {
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: This `fetch` call might fail, so we should have some error handling for that.

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                  }
+                </td>
+              </tr>
+              </tbody>
+            </table>
+          </ModalBody>
+          <ModalFooter>
+            <button type={'button'} className='btn btn-outline'
+                    onClick={() => this.close()}>Cancel</button>
+            <button id={'change_email'} type={'button'}
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: FYI, you don't need braces if the value is a simple string, e.g. `id='change_email' type='button'`

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                    className={'btn btn-primary'}
+                    onClick={() => this.updateAndSend()}>Apply</button>
+          </ModalFooter>
+        </Modal>}
+      {this.props.resend &&
+      <Modal id={'resend-instructions>'}>
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: That angle bracket in the id looks unintentional.

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        <ModalTitle>Resend Instructions</ModalTitle>
+        <ModalFooter>
+          <button type={'button'} className='btn btn-outline'
+                  onClick={() => this.close()}>Cancel</button>
+          <button type={'button'} id={'resend_instructions'}
+                  className={'btn btn-primary'}
+                  onClick={() => this.send()}>Send</button>
+        </ModalFooter>
+      </Modal>}
+    </React.Fragment>;
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: Bundling the two modals together works ok, but generally I think it would be preferably to split these out into separate components. Particularly because they're independent flows, and they have their own state.

File: ui/src/app/views/account-creation-modals/component.tsx [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    };
+    const args: FetchArgs = ProfileApiFetchParamCreator().resendWelcomeEmail(request);
+    fetch(this.fullUrl(args.url), args.options).then(() => {
+      this.setState({resending: false});
+    });
+    this.props.closeFunction();
+  }
+
+  leaveFocusEmail(evt): void {
+    this.setState({emailOffFocus: true, contactEmail: evt.target.value});
```
Comment (panentheos) [inline ui/src/app/views/account-creation-modals/component.tsx]: Rather than setting `contactEmail` here, it's more typical to use an `onChange` handler on the input, which will fire changes on every keystroke. You can still track the focus state for the purpose of toggling the error message.

File: ui/angular.json [hunk 10]
```diff
@@ -24,7 +24,8 @@
               "node_modules/@clr/icons/clr-icons.min.css",
               "node_modules/@clr/ui/clr-ui.min.css",
               "node_modules/nouislider/distribute/nouislider.min.css",
-              "src/styles.css"
+              "src/styles.css",
+              "src/assets/scss/styles.scss"
```
Comment (panentheos) [inline ui/angular.json]: If we're going to be using scss, we should configure and use CSS Modules so we maintain local scoping of class names.

https://github.com/webpack-contrib/css-loader#modules

This will probably require some tinkering with the webpack config, so it probably deserves its own PR, but we should tackle that before we accrue too many global class names.

Comment (UrsaStutsman) [reply ui/angular.json]: Yup - this was actually my exact worry with scss, because I couldn't figure out the webpack config tinkering required. We will definitely be discussing further at our tech talk the day we come back from break. But if you have any ideas about how to get this setup correctly that would be great! (Feel free to commit to this branch, as I'm now treating this branch as the scss demo)

**Comment by UrsaStutsman** [comment]
Closing because we are not currently going in this direction.
